### PR TITLE
Feature: Ipopt callbacks

### PR DESCRIPTION
**Pull request by jack-champagne**
Referenced the following during implementation:
https://github.com/jump-dev/Ipopt.jl/tree/4c156461ef1fda3c9f015520197afda4e8ca3e26?tab=readme-ov-file#solver-specific-callbacks
https://coin-or.github.io/Ipopt/IpStdCInterface_8h.html#ab86b55545fa3e0232a0e646409e0e82b

Add basic functionality, need to add:
- [x] test
- [x] docs
- [x] example

using internal callbacks, here are the primal variables as Ipopt optimizes. This was generated via saving the plots on the "Callback can get internal history" test

animation

### CODE REVIEW

File: src/problem_solvers.jl [hunk 1]
```diff
@@ -9,6 +9,7 @@ using ..Options
 
 using NamedTrajectories
 using MathOptInterface
+using Ipopt
 const MOI = MathOptInterface
```
Comment (aarontrowbridge) [inline src/problem_solvers.jl:14]: would be nice to add docstring for solve :)

**Comment by aarontrowbridge** [comment]
could we add this gif to the docs somewhere

File: src/problem_solvers.jl (hunk 1 above)
Comment (jack-champagne) [reply src/problem_solvers.jl:14]: Definitely, just noting it here - adding some features to save best trajectory (on various parameters), access to variables saved inside of callbacks programmatically (`get_callbacks`)

File: test/test_utils.jl [hunk 2]
```diff
@@ -116,3 +116,21 @@ function named_trajectory_type_1(; free_time=false)
...
+    ψ_target = ComplexF64[0.0, 1.0]
+    prob = QuantumStateSmoothPulseProblem(
+        sys, ψ_init, ψ_target, T, Δt;
+        ipopt_options=IpoptOptions(print_level=1), 
+        piccolo_options=PiccoloOptions(verbose=false)
+    )
+    if return_system
+        return prob, sys
+    else
+        return prob
```
Comment (jack-champagne) [inline test/test_utils.jl:134]: Thx @andgoldschmidt !

File: src/callbacks.jl [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+export make_save_best_trajectory_callback
+
+using NamedTrajectories
+
+import ..QuantumControlProblem
+import ..Problems: get_datavec
+
+function make_save_best_trajectory_callback(prob::QuantumControlProblem, traj::NamedTrajectory)
+    best_traj = NamedTrajectory(get_datavec(prob), prob.trajectory)
```
[the remark this answers is not in the document]
Comment (jack-champagne) [reply src/callbacks.jl]: This should be done correctly now

**Pull request merged**